src/NLog/Internal/PropertyHelper.cs

```diff
@@ -365,13 +362,14 @@ private static bool TryCreateCollectionObject(Type collectionType, string valueR
             //not checking "implements" interface as we are creating HashSet<T> or List<T> and also those checks are expensive
             if (isSet || typeDefinition == typeof(List<>) || typeDefinition == typeof(IList<>) || typeDefinition == typeof(IEnumerable<>)) //set or list/array etc
             {
+                object[] constructorArgs = isSet ? CreateHashSetConstructorArgs(obj, propInfo) : ArrayHelper.Empty<object>();
+
                 //note: type.GenericTypeArguments is .NET 4.5+ 
                 collectionItemType = collectionType.GetGenericArguments()[0];
-
                 var listType = isSet ? typeof(HashSet<>) : typeof(List<>);
                 var genericArgs = collectionItemType;
                 var concreteType = listType.MakeGenericType(genericArgs);
-                collectionObject = Activator.CreateInstance(concreteType);
+                collectionObject = constructorArgs?.Length > 0 ? Activator.CreateInstance(concreteType, constructorArgs) : Activator.CreateInstance(concreteType);
```

Comment:
Is this safe for all types? I think we should search for the proper ctor with reflection (and fallback to default ctor)

Reply:
With this?

https://docs.microsoft.com/en-us/dotnet/api/system.type.getconstructor?view=netframework-4.8#System_Type_GetConstructor_System_Type___

Another option is a try-catch

Reply:
The code for handling generic stuff is annoying, and this works.

Reply:
But please push your changes.

Reply:
Now calling GetConstructor pretty sure the code path is the same as Activator.CreateInstance, but now a little more convoluted :)

Reply:
👍